ceno_zkvm/src/e2e.rs

```diff
@@ -219,38 +142,257 @@ pub fn run_e2e_gen_witness<E: ExtensionField, PCS: PolynomialCommitmentScheme<E>
+            &emul_result.final_mem_state.mem,
+            &emul_result
+                .final_mem_state
+                .io
+                .iter()
+                .map(|rec| rec.cycle)
+                .collect_vec(),
+            &emul_result.final_mem_state.priv_io,
         )
         .unwrap();
     // assign program circuit
     zkvm_witness
-        .assign_table_circuit::<ProgramTableCircuit<E>>(&zkvm_cs, &prog_config, vm.program())
+        .assign_table_circuit::<ProgramTableCircuit<E>>(
+            &system_config.zkvm_cs,
+            &system_config.prog_config,
+            program,
+        )
         .unwrap();
 
+    zkvm_witness
+}
+
+pub enum Checkpoint {
+    PreProving,
+    PreVerifying,
+    Complete,
+}
+
+pub type ProvingArgs<E, PCS> = (ZKVMProver<E, PCS>, ZKVMWitnesses<E>, PublicValues<u32>);
+pub type VerifyingArgs<E, PCS> = (ZKVMProof<E, PCS>, ZKVMVerifier<E, PCS>, Option<u32>);
+
+pub enum CheckpointOutput<E: ExtensionField, PCS: PolynomialCommitmentScheme<E>> {
+    PreProving(ProvingArgs<E, PCS>),
+    PreVerifying(VerifyingArgs<E, PCS>),
+    Complete,
+}
+
+pub fn run_e2e_with_checkpoint<E: ExtensionField, PCS: PolynomialCommitmentScheme<E>>(
+    program: Program,
+    platform: Platform,
+    stack_size: u32,
+    heap_size: u32,
+    hints: Vec<u32>,
+    max_steps: usize,
+    checkpoint: Checkpoint,
+) -> CheckpointOutput<E, PCS> {
+    // Detect heap as starting after program data.
+    let heap_start = program.image.keys().max().unwrap() + WORD_SIZE as u32;
+    let heap_addrs = heap_start..heap_start + heap_size;
+    let mut mem_padder = MemPadder::new(heap_addrs.end..platform.ram.end);
+    let mem_init = init_mem(&program, &platform, &mut mem_padder, stack_size, heap_size);
+
+    let program_params = ProgramParams {
+        platform: platform.clone(),
+        program_size: program.instructions.len(),
+        static_memory_len: mem_init.len(),
+        ..ProgramParams::default()
+    };
+
+    let system_config = construct_configs::<E>(program_params);
+
+    // IO is not used in this program, but it must have a particular size at the moment.
+    let io_init = mem_padder.padded_sorted(system_config.mmu_config.public_io_len(), vec![]);
+    let reg_init = system_config.mmu_config.initial_registers();
+
+    let init_full_mem = InitMemState {
+        mem: mem_init,
+        reg: reg_init,
+        io: io_init,
+        priv_io: vec![],
+    };
+
+    // Generate fixed traces
+    let zkvm_fixed_traces = generate_fixed_traces(&system_config, &init_full_mem, &program);
+
+    // Keygen
+    let pcs_param = PCS::setup(1 << MAX_NUM_VARIABLES).expect("Basefold PCS setup");
+    let (pp, vp) = PCS::trim(pcs_param, 1 << MAX_NUM_VARIABLES).expect("Basefold trim");
+    let pk = system_config
+        .zkvm_cs
+        .clone()
+        .key_gen::<PCS>(pp.clone(), vp.clone(), zkvm_fixed_traces.clone())
+        .expect("keygen failed");
+    let vk = pk.get_vk();
+
```

Comment:
checkpoint 
```rust
let prover = ZKVMProver::new(pk);
let verifier = ZKVMVerifier::new(vk);
if let Checkpoint::PreProving = checkpoint {
    return CheckpointOutput::PreProving((prover, verifier, pi, exit_code));
}
```
proper place is here.

Then
emulate_program + generate_witness + run_e2e_proof 
those steps all belong to proving flow